soroban-simulation/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+    }
+    host.set_ledger_info(ledger_info.clone())
+        .context("cannot set ledger info")?;
+    host.set_base_prng_seed(rand::Rng::gen(&mut rand::thread_rng()))
+        .context("cannot set base prng seed")?;
+
+    // We make an assumption here:
+    // - if a transaction doesn't include any soroban authorization entries the client either
+    // doesn't know the authorization entries, or there are none. In either case it is best to
+    // record the authorization entries and return them to the client.
+    // - if a transaction *does* include soroban authorization entries, then the client *already*
+    // knows the needed entries, so we should try them in enforcing mode so that we can validate
+    // them, and return the correct fees and footprint.
+    let needs_auth_recording = invoke_hf_op.auth.is_empty();
+    if needs_auth_recording {
+        host.switch_to_recording_auth(true)
+            .context("cannot switch auth to recording mode")?;
+    } else {
+        host.set_authorization_entries(invoke_hf_op.auth.to_vec())
+            .context("cannot set authorization entries")?;
+    }
+
+    // Run the preflight.
+    let maybe_result = host
+        .invoke_function(invoke_hf_op.host_function.clone())
+        .context("host invocation failed");
+    let auths: VecM<SorobanAuthorizationEntry> = if needs_auth_recording {
+        let payloads = host.get_recorded_auth_payloads()?;
+        VecM::try_from(
+            payloads
+                .iter()
+                .map(recorded_auth_payload_to_xdr)
+                .collect::<Result<Vec<_>>>()?,
+        )?
+    } else {
+        invoke_hf_op.auth
+    };
+
+    let budget = host.budget_cloned();
+    // Recover, convert and return the storage footprint and other values to C.
+    let (storage, events) = host.try_finish().context("cannot finish host invocation")?;
+
+    let diagnostic_events = host_events_to_diagnostic_events(&events);
+    let result = match maybe_result {
+        Ok(r) => r,
+        // If the invocation failed, try to at least add the diagnostic events
+        Err(e) => {
+            return Ok(PreflightResult {
+                // See https://docs.rs/anyhow/latest/anyhow/struct.Error.html#display-representations
+                error: format!("{e:?}"),
+                events: diagnostic_events,
+                ..Default::default()
+            });
+        }
+    };
+
+    let invoke_host_function_with_auth = InvokeHostFunctionOp {
+        host_function: invoke_hf_op.host_function,
+        auth: auths.clone(),
+    };
+    let (transaction_data, min_fee) = fees::compute_host_function_transaction_data_and_min_fee(
+        &invoke_host_function_with_auth,
+        &ledger_storage_rc,
+        &storage,
+        &budget,
+        &diagnostic_events,
+        &result,
+        bucket_list_size,
+        ledger_info.sequence_number,
+    )
+    .context("cannot compute resources and fees")?;
+
+    let restore_preamble = compute_restore_preamble(
```

Comment:
I would still prefer to decouple the restore preamble logic from this function in this PR and switch to using the `SnapshotSource` trait instead of `LedgerStorage`. I would do that sooner or later, but if we do that now, we'll be able to preserve the interface of this function. OTOH I might need to change the interface more, so maybe it doesn't matter in the end. Up to you.

Reply:
Let's do it later if you are OK with it.